File: frontend/src/desktop/services/conversations.ts
import { invokeDesktop } from "./tauri";

export type DesktopConversation = {
  id: string;
  title: string;
  created_at: string;
  updated_at: string;
  pinned: boolean;
  archived: boolean;
};

export type DesktopMessage = {
  id: string;
  conversation_id: string;
  role: "user" | "agent" | "assistant" | "system";
  content: string;
  response_json?: string | null;
  created_at: string;
};

export type DesktopHistoryHit = {
  conversation_id: string;
  conversation_title: string;
  message_id: string;
  role: DesktopMessage["role"];
  content: string;
  created_at: string;
  score: number;
  snippet: string;
};

export type DesktopConversationSnapshotMessage = {
  role: DesktopMessage["role"];
  content: string;
  response_json?: string | null;
};

export function listConversations() {
  return invokeDesktop<DesktopConversation[]>("list_conversations");
}

export function listArchivedConversations() {
  return invokeDesktop<DesktopConversation[]>("list_archived_conversations");
}

export function createConversation(title: string) {
  return invokeDesktop<DesktopConversation>("create_conversation", { title });
}

export function updateConversationTitle(conversationId: string, title: string) {
  return invokeDesktop<void>("update_conversation_title", { conversationId, title });
}

export function updateConversationPinned(conversationId: string, pinned: boolean) {
  return invokeDesktop<void>("update_conversation_pinned", { conversationId, pinned });
}

export function archiveConversation(conversationId: string) {
  return invokeDesktop<void>("archive_conversation", { conversationId });
}

export function restoreConversation(conversationId: string) {
  return invokeDesktop<void>("restore_conversation", { conversationId });
}

export function listMessages(conversationId: string) {
  return invokeDesktop<DesktopMessage[]>("list_messages", { conversationId });
}

export function searchHistory(query: string, conversationId?: string, excludeCurrent = false, limit = 8) {
  return invokeDesktop<DesktopHistoryHit[]>("search_history", {
    query,
    conversationId,
    excludeCurrent,
    limit,
  });
}

export function appendMessage(
  conversationId: string,
  role: DesktopMessage["role"],
  content: string,
  responseJson?: string,
) {
  return invokeDesktop<DesktopMessage>("append_message", {
    conversationId,
    role,
    content,
    responseJson,
  });
}

export function saveConversationSnapshot(
  conversationId: string,
  title: string,
  messages: DesktopConversationSnapshotMessage[],
) {
  return invokeDesktop<void>("save_conversation_snapshot", {
    conversationId,
    title,
    messages,
  });
}

export function replaceMessageAfterEdit(messageId: string, content: string) {
  return invokeDesktop<void>("replace_message_after_edit", { messageId, content });
}

export function truncateConversationAfterMessage(messageId: string) {
  return invokeDesktop<void>("truncate_conversation_after_message", { messageId });
}

export function forkConversationFromMessage(messageId: string) {
  return invokeDesktop<DesktopConversation>("fork_conversation_from_message", { messageId });
}

File: src-tauri/src/app/storage_commands/conversations.rs
use super::logic::with_session;
use crate::db::DbState;
use crate::models::{Conversation, HistoryHit, Message};

#[tauri::command]
pub fn list_conversations(db: tauri::State<DbState>) -> Result<Vec<Conversation>, String> {
    with_session(&db, |session| session.list_conversations(false))
}

#[tauri::command]
pub fn list_archived_conversations(db: tauri::State<DbState>) -> Result<Vec<Conversation>, String> {
    with_session(&db, |session| session.list_conversations(true))
}

#[tauri::command]
pub fn create_conversation(
    db: tauri::State<DbState>,
    title: String,
) -> Result<Conversation, String> {
    with_session(&db, |session| session.create_conversation(&title))
}

#[tauri::command]
pub fn update_conversation_title(
    db: tauri::State<DbState>,
    conversation_id: String,
    title: String,
) -> Result<(), String> {
    with_session(&db, |session| {
        session.rename_conversation(&conversation_id, &title)
    })
}

#[tauri::command]
pub fn update_conversation_pinned(
    db: tauri::State<DbState>,
    conversation_id: String,
    pinned: bool,
) -> Result<(), String> {
    with_session(&db, |session| {
        session.set_conversation_pinned(&conversation_id, pinned)
    })
}

#[tauri::command]
pub fn archive_conversation(
    db: tauri::State<DbState>,
    conversation_id: String,
) -> Result<(), String> {
    with_session(&db, |session| {
        session.archive_conversation(&conversation_id)
    })
}

#[tauri::command]
pub fn restore_conversation(
    db: tauri::State<DbState>,
    conversation_id: String,
) -> Result<(), String> {
    with_session(&db, |session| {
        session.restore_conversation(&conversation_id)
    })
}

#[tauri::command]
pub fn delete_conversation_local(
    db: tauri::State<DbState>,
    conversation_id: String,
) -> Result<(), String> {
    with_session(&db, |session| session.delete_conversation(&conversation_id))
}

#[tauri::command]
pub fn list_messages(
    db: tauri::State<DbState>,
    conversation_id: String,
) -> Result<Vec<Message>, String> {
    with_session(&db, |session| session.list_messages(&conversation_id))
}

#[tauri::command]
pub fn search_history(
    db: tauri::State<DbState>,
    query: String,
    conversation_id: Option<String>,
    exclude_current: Option<bool>,
    limit: Option<usize>,
) -> Result<Vec<HistoryHit>, String> {
    with_session(&db, |session| {
        session.search_history(
            &query,
            conversation_id.as_deref(),
            exclude_current.unwrap_or(false),
            limit.unwrap_or(8),
        )
    })
}

#[tauri::command]
pub fn append_message(
    db: tauri::State<DbState>,
    conversation_id: String,
    role: String,
    content: String,
    response_json: Option<String>,
) -> Result<Message, String> {
    with_session(&db, |session| {
        session.append_message(&conversation_id, &role, &content, response_json)
    })
}
